proxy/src/coco/peer.rs

```diff
@@ -0,0 +1,617 @@
+        return Err(error::Error::EntityExists(urn));
+    } else {
+        let _repo = storage.create_repo(&user)?;
+    }
+
+    Ok(user)
+}
+
+/// Verify a user using a fake resolver that resolves the user to itself.
+///
+/// # Errors
+///
+/// If any of the verification steps fail
+pub async fn verify_user(user: user::User<entity::Draft>) -> Result<User, error::Error> {
+    let fake_resolver = FakeUserResolver(user.clone());
+    let verified_user = user
+        .check_history_status(&fake_resolver, &fake_resolver)
+        .await?;
+    Ok(verified_user)
+}
+
+/// Equips a repository with a rad remote for the given id. If the directory at the given path
+/// is not managed by git yet we initialise it first.
+fn setup_remote(
+    peer: &PeerApi,
+    path: impl AsRef<std::path::Path>,
+    id: &librad::hash::Hash,
+    default_branch: &str,
+) -> Result<(), error::Error> {
+    // Check if directory at path is a git repo.
+    if git2::Repository::open(&path).is_err() {
+        let repo = git2::Repository::init(&path)?;
+        // First use the config to initialize a commit signature for the user.
+        let sig = repo.signature()?;
+        // Now let's create an empty tree for this commit
+        let tree_id = {
+            let mut index = repo.index()?;
+
+            // For our purposes, we'll leave the index empty for now.
+            index.write_tree()?
+        };
+        let tree = repo.find_tree(tree_id)?;
+        // Normally creating a commit would involve looking up the current HEAD
+        // commit and making that be the parent of the initial commit, but here this
+        // is the first commit so there will be no parent.
+        repo.commit(
+            Some(&format!("refs/heads/{}", default_branch)),
+            &sig,
+            &sig,
+            "Initial commit",
+            &tree,
+            &[],
+        )?;
+    }
+
+    let repo = git2::Repository::open(path)?;
+
+    if let Err(err) = repo.resolve_reference_from_short_name(default_branch) {
+        log::error!("error while trying to find default branch: {:?}", err);
+        return Err(error::Error::DefaultBranchMissing(
+            id.to_string(),
+            default_branch.to_string(),
+        ));
+    }
+
+    let monorepo = peer.paths().git_dir().join("");
+    let namespace_prefix = format!("refs/namespaces/{}/refs", id);
+    let mut remote = repo.remote_with_fetch(
+        "rad",
+        &format!(
+            "file://{}",
+            monorepo.to_str().expect("unable to get str for monorepo")
+        ),
+        &format!("+{}/heads/*:refs/heads/*", namespace_prefix),
+    )?;
+    repo.remote_add_push(
+        "rad",
+        &format!("+refs/heads/*:{}/heads/*", namespace_prefix),
+    )?;
+    remote.push(
+        &[&format!(
+            "refs/heads/{}:{}/heads/{}",
+            default_branch, namespace_prefix, default_branch
+        )],
+        None,
+    )?;
+
+    Ok(())
+}
+
+// TODO(xla): Move into coco::source.
```

Comment:
Any strong reason to not resolve this TODO in the context of this PR?

Reply:
Can do for both of these requests :ok_hand: :point_down:

Reply:
All three in fact